File: src/lib/storage.ts
/**
 * Load and save Automerge documents to disk.
 */

import * as Automerge from "#automerge";
import type { Project } from "./schema.js";

type Doc = Automerge.Doc<Project>;

/** Save an Automerge document to a file, then git-add it. */
export async function saveDoc(filePath: string, doc: Doc): Promise<void> {
  const binary = Automerge.save(doc);
  await Bun.write(filePath, binary);

  // Auto-stage so the file is included in the next commit
  Bun.spawnSync(["git", "add", filePath], {
    stdout: "ignore",
    stderr: "ignore",
  });
}

/** Load an Automerge document from a file. Returns null if file doesn't exist. */
export async function loadDoc(filePath: string): Promise<Doc | null> {
  const file = Bun.file(filePath);
  if (!(await file.exists())) return null;

  const buffer = await file.arrayBuffer();
  return Automerge.load<Project>(new Uint8Array(buffer));
}

File: src/rust/crates/agt-lib/src/storage.rs
//! Load and save Automerge documents to disk.

use anyhow::{Context, Result};
use automerge::AutoCommit;
use std::path::Path;

/// Save an Automerge document to a file, then git-add it.
pub fn save_doc(path: &Path, doc: &mut AutoCommit) -> Result<()> {
    let binary = doc.save();
    std::fs::write(path, binary)
        .with_context(|| format!("failed to write {}", path.display()))?;

    // Auto-stage so the file is included in the next commit
    let _ = std::process::Command::new("git")
        .args(["add", &path.to_string_lossy()])
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status();

    Ok(())
}

/// Load an Automerge document from a file. Returns None if file doesn't exist.
pub fn load_doc(path: &Path) -> Result<Option<AutoCommit>> {
    if !path.exists() {
        return Ok(None);
    }
    let bytes = std::fs::read(path)
        .with_context(|| format!("failed to read {}", path.display()))?;
    let doc = AutoCommit::load(&bytes)
        .with_context(|| format!("failed to load automerge doc from {}", path.display()))?;
    Ok(Some(doc))
}
